### ISSUE TITLE
Header and Footer Missing - ESI Block Caching

### ISSUE THREAD

**Issue by natedesigns** [nexcess/magento-turpentine#1149]
Hi, 

New to varnish and Turpentine - so please be gentle. I believe that I've exhausted all posts on this topic from other posts.

I'm running Magento 1.9.1.1, Varnish 4.02 and Turpentine latest. 

When running an initial page load, I'm getting the whole page. However on the second hit, header and footer have disappeared. This was happening with the default turpentine_esi.xml, so I've tried commenting out the header and footer blocks, but this hasn't changed anything. 

I've gone through troubleshooting. Turning error reporting on/off, making sure all the files are actually there as suggested in [Blank ESI request](https://github.com/nexcess/magento-turpentine/issues/285).

I have the log setup which is giving me:
`2016-04-04T11:45:36+00:00 DEBUG (7): TURPENTINE: Cleared ESI cache for client (4531e1f4f8487eecfbd056ae1e6b5d41) on event: core_session_abstract_add_message  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: root  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: head  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: head.calendar  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_notices  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: header  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: menu  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: notifications  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: notification_baseurl  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache_notifications  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: notification_survey  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: notification_security  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: sagepaysuite_index_notifications  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: notification_licensemanager  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: left  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: messages  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: content  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.child1  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.child2  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.grid  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.grid.child3  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_59  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_60  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_61  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_62
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_63
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_64
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_65  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_66  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_67  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_68  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.grid.child5
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.grid.child6  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: formkey  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_69  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.grid.child4  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cache.additional  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: page.cache.additional  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: kingcatch.image.flush  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: regenerate.navigation.btn  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: turpentine_management  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: formkey  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: footer  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: js  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: profiler  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: before_body_end  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: customgrid.column_collection_renderers_descriptions  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: customgrid.messages  
2016-04-04T11:45:38+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: before_body_end.child2  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: root  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: head  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: head.theme.scripts  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: js_cookies  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: optional_zip_countries  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: google_analytics  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: awafptc.init  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: abtester  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: head.theme.css  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: after_body_start  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: turpentine_notices  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_notices  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: header  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: header  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: main_before_top  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: breadcrumbs  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: preface  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ultraslideshow  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_122  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: i_block_home_preface  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: block_home_preface  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_messages  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: global_messages  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: content  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: page_content_heading  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cms.wrapper  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cms_page  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_123  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_124  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_125  
2016-04-04T11:45:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_126  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_129  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_130  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_131  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_132  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_133  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_130  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_134  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_130  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_135  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_136  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_137  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_138  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_139  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_140  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_128  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_141  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_130  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_142  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_143  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: messages  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: messages  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: postscript  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: i_block_home_postscript  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: block_home_postscript  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: footer  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: footer  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: before_body_end  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: footer.theme.scripts  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_cookie_notice  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: global_cookie_notice  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: pap_clicktracking  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: awafptc.popup.container  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ultramenuoerlay  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: zendesk_support_tab  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: pageendscript  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: pageendscript.theme.scripts  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_145  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_146  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ANONYMOUS_144  
2016-04-04T11:45:46+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: core_profiler  
2016-04-04T11:45:46+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1  
2016-04-04T11:45:46+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 1  
2016-04-04T11:45:47+00:00 DEBUG (7): TURPENTINE: Generated form_key: VW3PLxJbYv4XiL8B  
2016-04-04T11:45:47+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1  
2016-04-04T11:45:47+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 0  
2016-04-04T11:45:48+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: awafptc.popup  
2016-04-04T11:45:48+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1  
2016-04-04T11:45:48+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 0`  

To me, the last two lines seem linked, but I'm not sure at the moment what's causing this. 

I have flushed cache (storage and Magento) in between any changes. 

Looking in my system.log I only have warnings from other extensions. And no related exceptions in exceptions.log

I am not using mod_pagespeed like [link](https://github.com/nexcess/magento-turpentine/issues/37)
I have got symlinks on as per fix [here](https://github.com/nexcess/magento-turpentine/issues/496)

I am getting the following headers returned on my homepage .

**First page request - (header/footer present)**
`Request URL:http://############/  
Request Method:GET  
Status Code:200 OK  
Remote Address:5.77.62.165:80  
Response Headers  
view source  
Cache-Control:no-store, no-cache, must-revalidate, post-check=0, pre-check=0  
Connection:close  
Content-Encoding:gzip  
Content-Type:text/html; charset=UTF-8  
Date:Mon, 04 Apr 2016 11:56:21 GMT  
Expires:Thu, 19 Nov 1981 08:52:00 GMT  
Pragma:no-cache  
Server:nginx  
Set-Cookie:frontend=2ec267116f72391c7027de001ffa3cd8; expires=Tue, 05-Apr-2016 11:56:20 GMT;   Max-Age=86400; path=/; domain=/############  
Transfer-Encoding:chunked  
Vary:Accept-Encoding  
X-Cache:HIT from Backend  
X-Frame-Options:SAMEORIGIN  
X-Powered-By:PHP/5.5.32  
Request Headers  
view source  
Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8  
Accept-Encoding:gzip, deflate, sdch  
Accept-Language:en-US,en;q=0.8  
Cache-Control:no-cache  
Connection:keep-alive  
Cookie:adminhtml=bd2c3874f16a40bf9e58f8be80f993cf  
Host:############  
Pragma:no-cache  
Upgrade-Insecure-Requests:1  
User-Agent:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.110 Safari/537.36`

**Second page request (no header/footer)**:
`Request URL:http://######/  
Request Method:GET  
Status Code:200 OK  
Remote Address:##  
Response Headers  
view source  
Age:459  
Connection:keep-alive  
Content-Encoding:gzip  
Content-Type:text/html; charset=UTF-8  
Date:Mon, 04 Apr 2016 11:45:46 GMT  
Server:nginx  
Transfer-Encoding:chunked  
Via:1.1 varnish-v4  
X-Cache:HIT from Backend  
X-Frame-Options:SAMEORIGIN  
X-Powered-By:PHP/5.5.32  
X-Turpentine-Cache:1  
X-Turpentine-Esi:1  
X-Varnish:327814 819274  
X-Varnish-Currency:  
X-Varnish-Esi-Access:  
X-Varnish-Esi-Method:  
X-Varnish-Hits:2  
X-Varnish-Host:#########  
X-Varnish-Set-Cookie:frontend=4531e1f4f8487eecfbd056ae1e6b5d41; expires=Tue, 05-Apr-2016   11:45:45 GMT; Max-Age=86400; path=/; domain=#########  
X-Varnish-Store:  
X-Varnish-URL:/  
Request Headers  
view source  
Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8  
Accept-Encoding:gzip, deflate, sdch  
Accept-Language:en-US,en;q=0.8  
Cache-Control:no-cache  
Connection:keep-alive  
Cookie:frontend=4531e1f4f8487eecfbd056ae1e6b5d41;   __cfduid=d5a51c6b79b6ecde90dc8ec6b780cc4361459765866; _hjIncludedInSample=1; PAPVisitorId=fe3ca38fdae83724685ec781610Hyz3G;   _vwo_uuid_v2=FCAC42FB3EFF3747E70F80530AC0359C|650d74ec51695eaeb4e62d9a9a1b610f; external_no_cache=1; __atuvc=1%7C14; __utmt=1;   adminhtml=9262ac2f5737378404918efef701cd43; _ga=GA1.5.359575813.1459765867; __utma=269279860.359575813.1459765867.1459765867.1459765867.1; __utmb=269279860.48.9.1459770797077; __utmc=269279860; __utmz=269279860.1459765867.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); __zlcmid=ZzfRzykbYCb3uf
Host:########  
Pragma:no-cache  
Upgrade-Insecure-Requests:1  
User-Agent:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.110 Safari/537.36`

I am aware that in my installation I will need to setup a [local.xml](https://github.com/nexcess/magento-turpentine/wiki/ESI_Cache_Policy) for a caching policy but wanted to get this issue out of the way first before proceeding. 

Your help is very much appreciated.

**Comment by aricwatson** [comment]
First thought:

> I'm running Magento 1.9.1.1, Varnish 4.02 and Turpentine latest.

What version of Linux? There's a known issue with CentOS7 and Varnish 4.0.3 that sound similar... I'd suggest trying either Varnish 4.1.x or going back to Varnish 4.0.3 and see if that helps? 

My second thought - are you running any other caching extensions? Can you try off *all* other cache types on the *Cache Management* screen other than the two Turpentine cache types?

**Comment by natedesigns** [comment]
CentOS7 6.7 - I had seen the post around CentOS7. But yes perhaps upgrading might help. I will do that after some more troubleshooting.

We are not running other caching types, I did just try disabling all other magento cache stores, other than Turpentines, but it didn't have an effect.

I have edited the turpentine_esi.xml taking everything out. This allowed the header and footer to return consistently. I then put back only the `<reference name="header">` section. The problem returned for the header. I'm now looking in this direction as to why this block is causing the error.
